test/e2e/instrumentation/tracing/apiserver.go

```diff
@@ -0,0 +1,201 @@
+		framework.ExpectNoError(err)
+
+		_, err = c.CoreV1().ConfigMaps(f.Namespace.Name).Create(context.Background(), opentelemetryConfigmap(), metav1.CreateOptions{})
+		framework.ExpectNoError(err)
+
+		ready := e2epod.CheckPodsRunningReady(f.ClientSet, f.Namespace.Name, []string{podName}, 20*time.Second)
+		framework.ExpectEqual(ready, true)
+	})
+
+	ginkgo.It("should send a request with a sampled trace context, and observe child spans from the collector pod", func() {
+		e2eskipper.SkipUnlessSSHKeyPresent()
+
+		ginkgo.By("Setting up OpenTelemetry.")
+		tp := sdktrace.NewTracerProvider(sdktrace.WithSampler(sdktrace.AlwaysSample()))
+		otel.SetTracerProvider(tp)
+		otel.SetTextMapPropagator(traces.Propagators())
+
+		ginkgo.By("Creating a context with a sampled parent span.")
+		// Note: After https://github.com/open-telemetry/opentelemetry-go/issues/2031
+		// is fixed, and we update our opentelemetry-go version, we won't be
+		// able to control sampling with parent spans anymore.  We will need to
+		// resolve https://github.com/kubernetes/kubernetes/issues/103186 in
+		// order to make this work again.
+		ctx, _ := tp.Tracer("apiservertest").Start(context.Background(), "OpenTelemetrySpan")
+
+		ginkgo.By(fmt.Sprintf("Checking for http target: %v in logs.", httpTarget))
+		gomega.Eventually(func() error {
+			// Send any request using the context with a sampled span
+			_, err := c.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+
+			// Get logs from the opentelemetry collector pod on the master.
+			// We use SSH because we can't fetch logs from master pods using
+			// the pod logs subresource.
+			result, err := e2essh.SSH(
+				fmt.Sprintf("sudo cat /var/log/pods/%v_%v_%v/%v/*", f.Namespace.Name, podName, otelPod.UID, containerName),
+				framework.APIAddress()+":22",
+				framework.TestContext.Provider,
+			)
+			if err != nil {
+				return err
+			}
+			if result.Stderr != "" {
+				return fmt.Errorf("non-empty stderr when querying for logs on the master: %v", result.Stderr)
+			}
+			logs := result.Stdout
+			// Check the opentelemetry collector logs to see if they contain
+			// our request. Since the apiserver has a ParentBased(NeverSample)
+			// sampler, it will only generate spans for our request with
+			// sampled parent above, and not for other requests.
+			if strings.Contains(logs, httpTarget) {
+				return nil
+			}
+			return fmt.Errorf("failed to find trace ID %v in log: \n%v", httpTarget, logs)
+		}, 1*time.Minute, 10*time.Second).Should(gomega.BeNil())
+
+	})
+})
+
+// masterNodeName fetches the first master node if there are multiple.
+// This won't work well for clusters with more than one master node.
+func masterNodeName(c clientset.Interface) (string, error) {
+	nodes, err := c.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+	framework.ExpectNoError(err)
+	for _, node := range nodes.Items {
+		if strings.HasSuffix(node.Name, "master") {
```

Comment:
this isn't a reliable way to find a master node is it? cc @smarterclayton

Reply:
We use it in test/e2e/framework/kubelet/stats.go to get metrics from control-plane components.

Reply:
that's an unreliable approach that should be eliminated, not propagated. xref https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/1143-node-role-labels/README.md#kubernetes-e2e-tests

Reply:
Ack.  Is there an alternative way to identify master nodes?

Reply:
It looks like some tests use `framework.APIAddress()` as the node name.  There is also `framework.TestContext.CloudConfig.MasterName`.  I'm going to try that approach

Reply:
Updated to use `framework.TestContext.CloudConfig.MasterName`, and to fall back to `framework.APIAddress()` if that isn't specified.